core/src/main/scala/io/finch/Endpoint.scala

```diff
@@ -82,7 +84,7 @@ trait Endpoint[A] { self =>
 
       final def apply(input: Input): Endpoint.Result[B] = self(input) match {
         case EndpointResult.Matched(rem, trc, out) =>
-          EndpointResult.Matched(rem, trc, out.flatMapF(this))
+          EndpointResult.Matched(rem, trc, out.run.flatMap(this))
```

Comment:
Maybe we shouldn't run it until the very last step (in `ToService`). I'd imagine, we can wrap user-defined function `A => Future[B]` wit `Task` and use just a normal `Task.flatMap`.

Also, I don't understand how it compiles. `out.run` returns `Future`, which is then flat-mapped onto a function. Not sure from where we get a new `Task` to pass into `EndpointMatched`.

Reply:
Oh I think the conversion happens implicitly: https://github.com/traneio/arrows/blob/master/arrows-twitter/src/main/scala/arrows/twitter/Arrow.scala#L619

We should probably avoid this bouncing and stay in a `Task` as long as possible.

Reply:
did some changes to remove the runs, it compiles and the implicit conversions are gone. Will try tomorrow to fix the test